src/backend/distributed/operations/shard_split.c

```diff
@@ -1145,3 +1221,650 @@ TryDropSplitShardsOnFailure(HTAB *mapOfShardToPlacementCreatedByWorkflow)
+		sourceColocatedShardIntervalList,
+		shardGroupSplitIntervalListList,
+		workersForPlacementList);
+
+	DropAllShardSplitLeftOvers(sourceShardToCopyNode, shardSplitHashMapForPublication);
+
+	MultiConnection *sourceConnection = GetNodeUserDatabaseConnection(connectionFlags,
+																	  sourceShardToCopyNode
+																	  ->
+																	  workerName,
+																	  sourceShardToCopyNode
+																	  ->
+																	  workerPort,
+																	  superUser,
+																	  databaseName);
+	ClaimConnectionExclusively(sourceConnection);
+
+	HTAB *mapOfShardToPlacementCreatedByWorkflow =
+		CreateEmptyMapForShardsCreatedByWorkflow();
+
+	/* Non-Blocking shard split workflow starts here */
+	PG_TRY();
+	{
+		/* 1) Physically create split children. */
+		CreateSplitShardsForShardGroup(mapOfShardToPlacementCreatedByWorkflow,
+									   shardGroupSplitIntervalListList,
+									   workersForPlacementList);
+
+		/*
+		 * 2) Create dummy shards due logical replication constraints.
+		 *    Refer to the comment section of 'CreateDummyShardsForShardGroup' for indepth
+		 *    information.
+		 */
+		CreateDummyShardsForShardGroup(
+			sourceColocatedShardIntervalList,
+			shardGroupSplitIntervalListList,
+			sourceShardToCopyNode,
+			workersForPlacementList);
+
+		/* 3) Create Publications. */
+		CreateShardSplitPublications(sourceConnection, shardSplitHashMapForPublication);
+
+		/*
+		 * 4) Create template replication Slot. It returns a snapshot. The snapshot remains
+		 * valid till the lifetime of the session that creates it. The connection is closed
+		 * at the end of the workflow.
+		 */
+		MultiConnection *templateSlotConnection = NULL;
+		char *snapShotName = CreateTemplateReplicationSlotAndReturnSnapshot(
+			shardIntervalToSplit, sourceShardToCopyNode, &templateSlotConnection);
+
+
+		/* 5) Do snapshotted Copy */
+		DoSplitCopy(sourceShardToCopyNode, sourceColocatedShardIntervalList,
+					shardGroupSplitIntervalListList, workersForPlacementList,
+					snapShotName);
+
+		/* 6) Execute 'worker_split_shard_replication_setup UDF */
+		List *replicationSlotInfoList = ExecuteSplitShardReplicationSetupUDF(
+			sourceShardToCopyNode,
+			sourceColocatedShardIntervalList,
+			shardGroupSplitIntervalListList,
+			workersForPlacementList);
+
+		/*
+		 * Subscriber flow starts from here.
+		 * Populate 'ShardSplitSubscriberMetadata' for subscription management.
+		 */
+		List *shardSplitSubscribersMetadataList =
+			PopulateShardSplitSubscriptionsMetadataList(
+				shardSplitHashMapForPublication, replicationSlotInfoList);
+
+		/* Create connections to the target nodes. TODO: can be refactored */
+		List *targetNodeConnectionList = CreateTargetNodeConnectionsForShardSplit(
+			shardSplitSubscribersMetadataList,
+			connectionFlags,
+			superUser, databaseName);
+
+		/* 7) Create copies of template replication slot */
+		char *templateSlotName = ShardSplitTemplateReplicationSlotName(
```

Comment:
I think given for considerable future, we plan to not support concurrent splits. The template slot name can be a constant which will simplify code a little.

Reply:
Sounds fine to me, although we might want to support that in the somewhat near future. But changing the naming scheme is pretty easy then.

Reply:
@niupre : I have kept it as debug info.